File: solutions/delimitersoup/DelimiterSoup.java
import java.util.Scanner;
import java.util.Stack;

public class DelimiterSoup {
    public static char getComplement(char l) {
        switch (l) {
            case '}': return '{';
            case  ')': return '(';
            case ']': return '[';
        }
         return '!';

    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        scanner.nextLine();
        String line = scanner.nextLine();
        Stack<Character> s = new Stack<>();
        for (int i = 0; i < n; i++) {
            char letter = line.charAt(i);
            if (letter == ' ')  continue;
            else if (getComplement(letter) == '!') s.add(letter);
            else if (s.isEmpty() || s.pop() != getComplement(letter)) {
                System.out.println(letter + " " + i);
                return;
            }
        }
        System.out.println("ok so far");
    }
}

File: delimitersoup/delimitersoup.py
n = int(input())
l = str(input())
s = []
d = {'{':'}', '[':']', '(':')'}
for i in range(n):
  if l[i] == ' ':
    continue
  if l[i] in d:
    s.append(d[l[i]])
  elif  len(s) == 0 or s.pop() != l[i]:
    print(l[i], i)
    exit()
print('ok so far') 